### DISCUSSION TITLE
[2023-05-22] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6797] [Weekly Demo Day]
I was mainly working on [Faster FRP engine implementation](https://github.com/enso-org/enso/issues/6795). I spend a lot of time on chatting with @Frizi and learning how his attempt at fixing this issue works (once again, Paweł, amazing job, it gave us amazing starting point and allowed to fast-forward many hypotheses ❤️). Unfortunately, I was feeling unwell a few days, so my progress is not as big as I wanted it to be, but here it is:

1. We have the initial implementation supporting events and behaviors, as the old one.
2. Currently, all data is passed by references, until it needs to be cached. This is a big improvement over the old implementation, where data was cloned on every FRP node. 
3. The new implementation allows defining nodes much easier than the old one. For example, `map2` implementation looks as follows (this is the FULL code):
   ```rust
    impl Network {
        pub fn map2<T0, T1: Default, Output: Data>(
            &self,
            n0: impl Node<Output = T0>,
            n1: impl Node<Output = T1>,
            f: impl Fn(&T0, &T1) -> Output + 'static,
        ) -> Stream<Output> {
            self.define_node((Event(n0), Behavior(n1)), move |event, t0, t1| event.emit(&f(t0, t1)))
        }
    }
    ```
4. The time of evaluating the new FRP is fast. It's approx 50-70x slower than just a loop with a function call (please remember, that FRP does much more logic than calling these functions), however, a simple function call is a good benchmark. For example, calling `100_000` times a 3 nodes FRP network, evaluates in approx 1ms on my computer.
    ```rust
    #[bench]
    fn bench_emit_plain_fn(bencher: &mut Bencher) {
        let mut sum = 0;
        bencher.iter(move || {
            for i in 0..REPS {
                sum += map_fn(&i, &i);
            }
            assert_ne!(sum, 0);
        });
    }

    #[bench]
    fn bench_emit_frp(bencher: &mut Bencher) {
        let net = Network::new();
        let src1 = net.source::<usize>();
        let src2 = net.source::<usize>();
        let m1 = net.map2(src1, src2, map_fn);
        src2.emit(2);

        bencher.iter(move || {
            let _keep = &net;
            for i in 0..REPS {
                src1.emit(i);
            }
        });
    }
    ```
5. The time of initializing FRP network is much faster as well. I am still working on optimizing it and I believe we can get much better times, but currently, the time needed to initialize `100_000` FRP nodes is 4ms on my machine, in comparison to 10ms in the old implementation.

**Comment by Procrat** [comment]
1. #6719. Picking a function from the CB that's defined in the main module now resolves to `Main.<func-name>` instead of `<project-name>.<func-name>`.
2. Spend some time re-layouting the top bar to address PR comments on #6474.

**Comment by JaroslavTulach** [comment]
Language server & engine performance improvement:

- Bug fix for flood of `suggestionsDatabaseUpdate` messages:  #6755

**Comment by MichaelMauderer** [comment]
* https://github.com/enso-org/enso/pull/6758: Fixed a performance regression introduced with one of the visualization bugs fixed
* Fixing a cluster of bugs around the initialization of visualization, which affect issues from #6770 and #6561 and  #6673  
  PR in progress and includes simplifications and unification of some logic of handling visualizations.

https://github.com/enso-org/enso/assets/1428930/3fcf4e78-260b-47f2-ad3b-f983017a2cb6

**Comment by kazcw** [comment]
1. Finished new edges, saving 10/42 (~24%) of the constantly-displayed draw calls (#6707).
2. Fixed a render latency that was causing flickering; the design change also reduces the frontend's CPU work by 5% (#6744).
3. Improved performance of handling frequent view-invalidating messages, saving 5s of CPU time with some backend versions (#6786).

**Comment by mwu-tow** [comment]
I am still working on Undo/Redo improvements (https://github.com/enso-org/enso/issues/6435).
Over the last week I focused in particular on the following issues:
* Metadata handling discrepancy between the GUI and the backend (https://github.com/enso-org/enso/pull/6718);
* Spurious transactions when opening the project;
* Empty edits.

**Comment by farmaazon** [comment]
### IDE now can recover from text synchronization errors.

Each time we get an error from applyEdit, the next change will be preceded with file reload from language server (with full invalidation to send all missing edits).
 
You can also force reload with  ctrl+shift+x shortcut. For each change after des

### In progress: The Column.from_vector will not be selected by default every time

https://github.com/enso-org/enso/assets/3919101/e9233997-02f3-4d15-997f-dadf6ee04141

In the above clip, you can see, that typing vector makes actual preview of _vector_ instead of random error before (as Column.from_vector was forcefully added). Also, after picking function and pressing space, I still see preview for the function. However, when further typing, the entries are filtered and best match is selected (and previewed).

As a rule, we select first entry in CB only when a) the input is empty b) we are filtering (so the user has some text on the left of the text cursor).
